add blog post: An introduction to use the domain mode of WildFly 35

liweinan opened the pull request.

liweinan commented:
I need to make more modifications to this blog post, but I'd like to create pull request here for review to see if this is proper to be published here.

liweinan commented:
I need to use WildFly 35 for this blog post.

jamezp commented:
@yersan Could you give this a review?

jamezp commented:
@liweinan Before this is merged, could you squash the commits?

yersan commented:
> @yersan Could you give this a review?

@jamezp yes, it is on my radar, I want to address it during this week

yersan reviewed (commented):
Thanks @liweinan , I've added some suggestions, mostly regarding the Domain Mode terminology since it can be easily confused by the reader.

liweinan commented:
@jamezp @yersan Thanks for the reviews! I'll update the text according to your comments :D

liweinan commented on _posts/2025-01-09-wildfly-domain-usage.adoc:
```diff
@@ -0,0 +1,394 @@
+
+Please note that I can’t use the IP address `127.0.0.1` here, because the WildFly server is listening to the public IP address of Machine A, which we have assigned to the `jboss.bind.address.management` property during the server startup process. Here is the screenshot of the admin page of the WildFly server:
+
+image:2025-01-wildfly-domain/02.png[image]
```
Unfortunately, my WiFi refreshes my machine's IP address automatically:

<img width="1280" alt="image" src="https://github.com/user-attachments/assets/232933ff-0914-47c3-9a46-fd621ff03acf" />

yersan replied:
I understand; updating it would force you to update all the other screen captures. In such a case, from my side feel free if you want to leave it as it is, that's ok too.

liweinan replied:
Okay!

liweinan commented:
@yersan I have updated the text and the pictures according to your comments. @jamezp I have squashed the commits.

Could you please help to review this PR again when you have time? Thanks!

yersan reviewed (commented):
Thanks @liweinan , this has good content. I added some additional comments related to language usage than the content. From my side, you have addressed the main points about Domain Mode terminology. The blog post has good hints and explains good concepts, e.g how to configure a managed server, the port offset concepts, how to configure security with Elytron. I found it very useful, so it LGTM regarding Domain Mode content.

@jamezp Could you take a look focussing on grammatical mistakes or expression usage? I think those should be better reviewed and approved by an English speaker.

liweinan commented on _posts/2025-01-31-wildfly-domain-usage.adoc:
```diff
@@ -0,0 +1,402 @@
+---
+layout: post
+title:  "An introduction to use the domain mode of WildFly 35"
```
I rewrite the title and the first paragraph to better explain the purpose of writing this blog post.

liweinan commented:
@yersan Thanks for the detailed review! I have updated the text to match your comments.

I'll wait for James to review it and then squash the commits.

jamezp commented on _posts/2025-01-31-wildfly-domain-usage.adoc:
```diff
@@ -0,0 +1,401 @@
+[source,xml]
+----
+<domain-controller>
+    <remote authentication-context="myCtx">
```
@yersan Are we not able to write this attribute with CLI? I'm not seeing the attribute defined, but I'm trying to make sure I'm not missing it.

yersan replied:
@jamezp yes, we can:

`/host=secondary:write-attribute(name=domain-controller.remote.authentication-context, value=myCtx)`

Good point about adding the CLI by the way

jamezp replied:
Ah, perfect. Thank you!

liweinan replied:
@yersan Is there a way to write the enclosing `discovery-options` options by the CLI command too?

liweinan replied:
I used this tool to dump my host config so I could fill the rest of the CLI commands:

- https://github.com/tfonteyn/profilecloner

Nevertheless, I encountered a problem using the tool to dump the host configuration of the secondary host. I raised this issue to the project:

- https://github.com/tfonteyn/profilecloner/issues/12

I'll add all the necessary CLI commands to this blog post and then squash the commits.

liweinan replied:
The tool is useful to me which helps me to learn to write the CLI commands :D

liweinan replied:
@yersan @jamezp With this command:

```
/host=secondary:write-attribute(name=domain-controller.remote.authentication-context, value=myCtx)
```

The generated config is:

```xml
    <domain-controller>
        <local/>
    </domain-controller>
```

Is there any way I can generate the config like this:

```xml
    <domain-controller>
        <remote authentication-context="myCtx">
            <discovery-options>
                <static-discovery name="primary" protocol="${jboss.domain.primary.protocol:remote+http}" host="${jboss.domain.primary.address}" port="${jboss.domain.primary.port:9990}"/>
            </discovery-options>
        </remote>
    </domain-controller>
```

liweinan replied:
According to the article here: https://www.wildfly.org/news/2017/10/10/Embedded-Host-Controller/

I write the command like this but seems it's not the correct way to do so:

```
[domain@embedded /] /host=secondary:write-remote-domain-controller(authentication-context="myCtx", protocol="${jboss.domain.primary.protocol:remote+http}", host="${jboss.domain.primary.address}", port="${jboss.domain.primary.port:9990}")
{
    "outcome" => "success",
    "result" => undefined,
    "response-headers" => {
        "warnings" => [{
            "warning" => "WFLYCTL0449: Operation write-remote-domain-controller against the resource at address /host=secondary is deprecated, and it might be removed in future
version. See the the output of the read-operation-description operation to learn more about the deprecation.",
            "level" => "WARNING",
            "operation" => {
                "address" => [("host" => "secondary")],
                "operation" => "write-remote-domain-controller"
            }
        }],
        "process-state" => "reload-required"
    },
    "server-groups" => undefined
}
```

jamezp replied:
I would not use this command. The config you have below is the default:
```
[domain@embedded /] /host=secondary/core-service=discovery-options/static-discovery=primary:read-resource
{
    "outcome" => "success",
    "result" => {
        "host" => expression "${jboss.domain.primary.address}",
        "port" => expression "${jboss.domain.primary.port:9990}",
        "protocol" => expression "${jboss.domain.primary.protocol:remote+http}"
    },
    "response-headers" => {"process-state" => "reload-required"}
}
```

That said, the path you'd use is `/host=secondary/core-service=discovery-options/static-discover=primary`.

yersan replied:
+1

@liweinan if you use the default host-secondary.xml you don't need to create these properties. Use the default ones as the base template and then modify them as you need. So, you would only have to add the authentication context as was described before:

`/host=secondary:write-attribute(name=domain-controller.remote.authentication-context, value=myCtx)`

If you need to modify the static-discovery ones, you can doit using the `write-attribute`:

`[domain@embedded /] /host=localhost/core-service=discovery-options/static-discovery=primary:write-attribute(name=
host  port  protocol`

liweinan replied:
Thanks for teaching @yersan

liweinan replied:
And this my worth checking: https://github.com/tfonteyn/profilecloner/issues/12#issuecomment-2654227956

jamezp commented on _posts/2025-01-31-wildfly-domain-usage.adoc:
```diff
@@ -0,0 +1,401 @@
+
+[source,xml]
+----
+<profile>
+    <subsystem xmlns="urn:jboss:domain:core-management:1.0"/>
+    <subsystem xmlns="urn:wildfly:elytron:community:18.0" final-providers="combined-providers"
+               disallowed-providers="OracleUcrypto" register-jaspi-factory="false">
+        <authentication-client>
+            <authentication-configuration sasl-mechanism-selector="DIGEST-MD5" name="myConfig"
+                                          authentication-name="admin" realm="ManagementRealm">
+                <credential-reference clear-text="123"/>
+            </authentication-configuration>
+            <authentication-context name="myCtx">
+                <match-rule match-host="${jboss.domain.primary.address}" authentication-configuration="myConfig"/>
+            </authentication-context>
+        </authentication-client>
+        ...
+    </subsystem>
+    ...
+</profile>
```
We should do this with CLI IMO. Probably before the step above.

```
/host=stalefish/subsystem=elytron/authentication-configuration=myConfig:add(sasl-mechanism-selector=DIGEST-MD5, authentication-name=admin, realm=ManagementRealm)
/host=stalefish/subsystem=elytron/authentication-context=myCtx:add(match-rules=[{match-host="${jboss.domain.primary.address}", authentication-configuration=myConfig}])
```

liweinan replied:
@jamezp Is there way to configure the password in the CLI command?

liweinan replied:
I tried out the command to add the config with credential:

```
[domain@embedded /] /host=secondary/subsystem=elytron/authentication-configuration=myConfig:add(sasl-mechanism-selector=DIGEST-MD5, authentication-name=admin, realm=ManagementRealm, credential-reference={clear-text="123"})
{"outcome" => "success"}
```

Generated config:

```xml
<authentication-client>
                <authentication-configuration name="myConfig" authentication-name="admin" realm="ManagementRealm" sasl-mechanism-selector="DIGEST-MD5">
                    <credential-reference clear-text="123"/>
                </authentication-configuration>
            </authentication-client>
```

liweinan replied:
I have updated the text accordingly. Thanks for teaching James!

jamezp commented on _posts/2025-01-31-wildfly-domain-usage.adoc:
```diff
@@ -0,0 +1,401 @@
+
+[source,xml]
+----
+<servers>
+    <!-- <server name="server-one" group="main-server-group"/> -->
+    <server name="server-two" group="other-server-group">
+        <socket-bindings port-offset="150"/>
+    </server>
+</servers>
```
Again with CLI:
```
/host=secondary/server-config=server-one:remove
```

jamezp reviewed (commented):
Just some minor comments about using CLI over editing the raw XML as this is generally preferred. You can do this with the embedded host controller by doing something like:

```bash
./bin/jboss-cli.sh
```

Then once in CLI:
```
embed-host-controller
```

You can add `--host-config=host-secondary.xml` to the `embed-host-controller` command to make changes to the secondary host.

liweinan commented:
Thanks for the detailed review James! I'll update the text accordingly.

liweinan commented on _posts/2025-01-31-wildfly-domain-usage.adoc:
```diff
@@ -0,0 +1,401 @@
+
+From the above command output, we can see the `admin` user is added and its password is `123`. This added user will be used on the Secondary Host Controller to connect to the Primary Host controller.
+
+Please note that we didn’t generate a secret value related to the generated `admin` user, which is already deprecated. So this kind of configuration on the Secondary Host Controller is deprecated:
+
+[source,xml]
+----
+<server-identities>
+  <secret value="..." />
+</server-identities>
+----
```
I'll rephrase this part.

liweinan commented on _posts/2025-01-31-wildfly-domain-usage.adoc:
```diff
@@ -0,0 +1,497 @@
+
+[source,xml]
+----
+<host xmlns="urn:jboss:domain:community:20.0" name="secondary">
```
@jamezp Is there a way to configure the host name here via CLI command?

liweinan replied:
@yersan @jamezp I tried to use the following command to edit the host name on the secondary host but failed:

```bash
[domain@embedded /] /host=current-host:write-attribute(name=name, value=secondary)
Failed to get the list of the operation properties: "WFLYDC0032: Operation read-operation-description for address [("host" => "current-host")] can only be handled by the Domain Controller; this host is not the Doma
in Controller"
```

I guess this means I can't directly edit the host name on the secondary host by using the embed server command am I correct?

liweinan replied:
The only difference I could see is that in your command it's: `/host=localhost`. I'll try with this value, thanks for checking!

liweinan replied:
After doing some experiments, I found the current host name affects the output and that's the root cause:

<img width="1904" alt="image" src="https://github.com/user-attachments/assets/87dc1b9f-cc7c-4371-9743-4c4f37f152fc" />

Now I understand how it works. Thanks for the help Yeray!

jamezp commented on _posts/2025-01-31-wildfly-domain-usage.adoc:
```diff
@@ -0,0 +1,523 @@
+
+[source,bash]
+----
+$ ./domain.sh --host-config=host.xml -Djboss.bind.address.management=192.168.0.115 -Djboss.bind.address=192.168.0.115 -Djboss.domain.primary.address=192.168.0.115
```
Not critical and this can stay, but the `host.xml` is the default value for the `--host-config` argument. Meaning in this specific case, it's not required.

liweinan replied:
add notes about the default value.

jamezp commented on _posts/2025-01-31-wildfly-domain-usage.adoc:
```diff
@@ -0,0 +1,523 @@
+}
+
+[domain@embedded /]
+[domain@embedded /] /host=secondary:reload(admin-only=true)
```
This would still be `/host=localhost` here.

jamezp commented on _posts/2025-01-31-wildfly-domain-usage.adoc:
```diff
@@ -0,0 +1,523 @@
+Here is the equivalent CLI command to do the configuration:
+
+[source,bash]
+----
+[domain@embedded /] /host=secondary/subsystem=elytron/authentication-configuration=myConfig:add(sasl-mechanism-selector=DIGEST-MD5, authentication-name=admin, realm=ManagementRealm, credential-reference={clear-text="123"})
+{"outcome" => "success"}
+----
+
+[source,bash]
+----
+[domain@embedded /] /host=secondary/subsystem=elytron/authentication-context=myCtx:add(match-rules=[{match-host="${jboss.domain.primary.address}", authentication-configuration=myConfig}])
+{"outcome" => "success"}
+----
```
These should be moved above the previous commands. We should create the `myCtx` resource before we assign it the host.

liweinan replied:
Okay! Moved.

liweinan commented:
@jamezp  Thanks for the review! I'll update the text accordingly and squash the commits.
